README.md

```diff
@@ -16,3 +16,127 @@ Why "None-aware attribute getter"?
+    "employee": {
+      "type": "object",
+      "properties": {
+        "name": {
+          "type": "string"
+        },
+        "payroll": {
+          "type": "object",
+          "properties": {
+            "frequency": {
+              "type": "string"
+            },
+            "salary": {
+              "type": "object",
+              "properties": {
+                "base": {
+                  "type": "object",
+                  "properties": {
+                    "amount": {
+                      "type": "number"
+                    },
+                    "currency": {
+                      "type": "string"
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      },
+      "required": [
+        "name"
+      ]
+    }
+  },
+  "required": [
+    "employee"
+  ]
+}
+```
+
+Examples of valid JSON response's payloads would be:
+
+```json
+{"employee": {"name": "Will Smith"}}
+{"employee": {"name": "Will Smith", "payroll":{"frequency": "monthly"}}}
+{"employee": {"name": "Will Smith", "payroll":{"frequency": "monthly", "salary": {"base": {"amount": 100000, "currency": "USD"}}}}}
+```
+
+Let's first assume that the JSON has been deserialized as nested `dict`s.
+This is a widespread manner of handling such data, as the `json` module from the CPython standard library provides the `json.loads` function that deserializes JSON objects to `dict`.
+In order to get, for instance, the amount of the employee's base salary, a common pattern is to chain `dict.get` and setting intermediate calls' default value to `{}`.
+
+```python
+amount: int | None = data.get("employee", {}).get("payroll", {}).get("salary", {}).get("base", {}).get("amount")
+```
+
+However, for various reasons, a common pattern is to parse the JSON payload into a Python object.
+
+For instance, using `pydantic`, one may define the following models.
+
+```python
+from pydantic import BaseModel
+
+
+class Employee(BaseModel):
+  name: str
+  payroll: Payroll | None
+
+
+class Payroll(BaseModel):
+  frequency: str
+  salary: Salary | None
+
+
+class Salary(BaseModel):
+  base: Base | None
+
+
+class Base(BaseModel):
+  amount: int | None
+  currency: str | None
+```
+
+Once the payload is parsed into such a model, it is no longer possible to use the same mechanism that we had with `dict.get`.
+
+Instead, code will look like something similar to:
+
+```python
+employee: Employee = ...
+amount = (
+    employee.payroll.salary.base.amount
+    if (
+      employee.payroll is not None
+      and employee.payroll.salary is not None
+      and employee.payroll.salary.base is not None
+    )
+    else None
+)
+```
+
+This results in verbose boilerplate code that arguably does not convey the intent.
+
+If Python were to have a None-aware attribute getter as we propose, the same result could be achieved with the code:
+
+```python
+employee: Employee = ...
+amount = employee?.payroll?.salary?.base
```

Comment:
[From Elis Byberi on discuss.python.org](https://discuss.python.org/t/introducing-a-safe-navigation-operator-in-python/35480/138):

> It’s very time-consuming and hard on the eyes to find every ? character in this line:
> ```python
> amount = employee?.payroll?.salary?.base
> ```
>
> Why don’t you just use try/except in a method or function?
> 
> ```python
> def get_amount(employee):
>     try:
>         return employee.payroll.salary.base
>     
>     except AttributeError:
>         return None
> ```

Reply:
In the [original post](https://discuss.python.org/t/introducing-a-safe-navigation-operator-in-python/35480/1) of this thread, the author mentioned what I think is a compelling argument:

> [It] helps in avoiding `AttributeError` exceptions caused by attempting to access attributes or methods on a NoneType object.

Suppose there used to be a 'bonus' field on the 'salary' object, but that has now changed. The `try: ... catch AttributeError: return None` methodology would now return `None` when it should, really, raise an AttributeError.

This example illustrates that there are two distinct semantics:
- I want a None returned when trying to access an attribute of a value (of type `T | None`) when that value is None.
- I want an AttributeError raised if I try to access a field on a non-None value and that attribute does not exist.

The `try: ... catch AttributeError: return None` does not allow to have these two distinct meanings.

On another note, this construct induces a lot of boilerplate code when this chaining happens in function arguments. For instance if I need to change representation:

```python
@dataclass
class CostUnit:
    frequency: str
    amount: int
    currency: str


# proposed
cost_unit = CostUnit(
    frequency=employee?.payroll?.salary?.frequency or "monthly",
    amount=employee?.payroll?.salary?.base?.amount or 0,
    currency=employee?.payroll?.salary?.currency or "USD",
)

# current
frequency: str
amount: int
currency: str
try:
    frequency = employee.payroll.salary.frequency
except AttributeError:
    frequency = "monthly"
try:
    amount = employee.payroll.salary.base.amount
except AttributeError:
    amount = 0
try:
    currency = employee.payroll.salary.currency
except AttributeError:
    currency = "USD"
cost_unit = CostUnit(
    frequency=frequency,
    amount=amount,
    currency=currency,
)
```

I would argue that, in this case, the proposed form conveys the intention better because it's more local.

Reply:
> I want an AttributeError raised if I try to access a field on a non-None value and that attribute does not exist.

How could this happen? Can you show me an example? Should you be using **assert** statements or static type checkers?

> used to be a ‘bonus’ field on the ‘salary’ object, but that has now changed

You should fix the code where this instance is used. Personally, I do use exceptions to aid in debugging.

You can use `AttributeError` in this case:
* I want a None returned when trying to access an attribute of a value (of type `T | None`) when that value is None.

I don't see two distinct meanings.